### PR TITLE
Delete routes when endpoint destroyed

### PR DESCRIPTION
**Pull request by plwhite**
Fixes issue #184.

### CODE REVIEW

File: calico/felix/endpoint.py [hunk 1]
```diff
@@ -92,8 +92,23 @@ def __init__(self, uuid, mac):
...
+        """
+        Delete a programmed endpoint. Remove the routes, then the rules.
+        """
+        for type in (futils.IPV4, futils.IPV6):
+            try:
+                if devices.tap_exists(self.tap):
+                    ips = devices.list_tap_ips(type, self.tap)
+                    for ip in ips:
+                        devices.del_route(type, ip, self.tap)
+            except:
```
Comment (Lukasa) [inline calico/felix/endpoint.py]: Let's avoid the naked `except`.

Comment (Lukasa) [inline calico/felix/endpoint.py]: This is also overbroad. We check if the tap exists in both branches, so we should be able to reduce the scope of this try...except block.

Comment (Lukasa) [inline calico/felix/endpoint.py]: Additionally, I think if the tap existed when we started and doesn't exist when we're done we don't need to rethrow. This logic is definitely included in the current code and we want it, but I think we can still get it with just one code block.

**Comment by Lukasa** [comment]
One note here. =)

**Comment by plwhite** [comment]
What we have now is this.

    for type in (futils.IPV4, futils.IPV6):
        try:
            if devices.tap_exists(self.tap):
                ips = devices.list_tap_ips(type, self.tap)
                for ip in ips:
                    devices.del_route(type, ip, self.tap)
        except:
            if devices.tap_exists(self.tap):
                raise
            break

I think you are proposing this :

    if devices.tap_exists(self.tap):
        for type in (futils.IPV4, futils.IPV6):
            try:
                ips = devices.list_tap_ips(type, self.tap)
                for ip in ips:
                    devices.del_route(type, ip, self.tap)
            except (futils.WhateverTheExceptionIs):
                if devices.tap_exists(self.tap):
                    raise
                break


Is that right? No objection in principle to that restructuring.

**Comment by Lukasa** [comment]
@plwhite Exactly right. I'm a big believer in small try...except blocks. =D

**Comment by plwhite** [comment]
Back to you. Note that I did a `push -f` so that the intermediate version would not appear in the history, but small enough that should not matter much.

**Comment by Lukasa** [comment]
Cool, LGTM. :cake:

**Pull request merged**